openmc/model/triso.py

```diff
@@ -153,3 +520,502 @@ def create_triso_lattice(trisos, lower_left, pitch, shape, background):
+    """Close random packing of particles using the Jodrey-Tory algorithm.
+
+    Parameters
+    ----------
+    domain : openmc.model._Domain
+        Container in which to pack particles.
+    particles : numpy.ndarray
+        Initial Cartesian coordinates of centers of particles.
+    contraction_rate : float
+        Contraction rate of outer diameter.
+
+    """
+
+    def add_rod(d, i, j):
+        """Add a new rod to the priority queue.
+
+        Parameters
+        ----------
+        d : float
+            distance between centers of particles i and j.
+        i, j : int
+            Index of particles in particles array.
+
+        """
+
+        rod = [d, i, j]
+        rods_map[i] = (j, rod)
+        rods_map[j] = (i, rod)
+        heappush(rods, rod)
+
+
+    def remove_rod(i):
+        """Mark the rod containing particle i as removed.
+
+        Parameters
+        ----------
+        i : int
+            Index of particle in particles array.
+
+        """
+
+        if i in rods_map:
+            j, rod = rods_map.pop(i)
+            del rods_map[j]
+            rod[1] = None
+            rod[2] = None
+
+
+    def pop_rod():
+        """Remove and return the shortest rod.
+
+        Returns
+        -------
+        d : float
+            distance between centers of particles i and j.
+        i, j : int
+            Index of particles in particles array.
+
+        """
+
+        while rods:
+            d, i, j = heappop(rods)
+            if i is not None and j is not None:
+                del rods_map[i]
+                del rods_map[j]
+                return d, i, j
+
+
+    def create_rod_list():
+        """Generate sorted list of rods (distances between particle centers).
+
+        Rods are arranged in a heap where each element contains the rod length
+        and the particle indices. A rod between particles p and q is only
+        included if the distance between p and q could not be changed by the
+        elimination of a greater overlap, i.e. q has no nearer neighbors than p.
+
+        A mapping of particle ids to rods is maintained in 'rods_map'. Each key
+        in the dict is the id of a particle that is in the rod list, and the
+        value is the id of its nearest neighbor and the rod that contains them.
+        The dict is used to find rods in the priority queue and to mark removed
+        rods so rods can be "removed" without breaking the heap structure
+        invariant.
+
+        """
+
+        # Create KD tree for quick nearest neighbor search
+        tree = scipy.spatial.cKDTree(particles)
+
+        # Find distance to nearest neighbor and index of nearest neighbor for
+        # all particles
+        d, n = tree.query(particles, k=2)
+        d = d[:,1]
+        n = n[:,1]
+
+        # Array of particle indices, indices of nearest neighbors, and
+        # distances to nearest neighbors
+        a = np.dstack(([i for i in range(len(n))], n, d))[0]
```

Comment:
That list could be `list(range(len(n)))` or `list(range(n.size))` or `np.arange(n.size)`.

Also, consider `np.vstack(...).T` which, at least to me, is a little more obvious (I think that does the same thing)